include/pybind11/stl_bind.h

```diff
@@ -694,6 +694,26 @@ struct ItemsViewImpl : public detail::items_view {
     Map &map;
 };
 
+template <typename KeyType>
+std::string format_message_key_error(const KeyType &k) {
+    std::string message;
```

Comment:
> What is the benefit of `set_error + throw error_already_set()` over `throw key_error()` ?

The main idea is to avoid a roundtrip: Conversion of a Python `str` to C++ `std::string` and then back to a Python `str`; this involves unicode encoding/decoding.

So here (line 699) I'd make this `str message;`

This function is going in a great direction, but I'd work on it more (quite possibly it can be reused in other contexts; later).

I'd split this into layers, to support this:

Only the `cast(k)` needs to be templated, and it should only run once (right now it may run twice).

I'd consider using a fallback message, maybe `"pybind11::bind_map key"` instead of the empty string.

Reply:
Done, PTAL.